go/services/connector/main.go

```diff
@@ -0,0 +1,620 @@
+}
+
+// FileMountMap tracks device endpoints and their associated assets.
+type FileMountMap struct {
+	mu             sync.RWMutex
+	fileMountPath  string
+	deviceMap      map[string]DeviceData
+	createDeviceTx chan DeviceNotification
+	logger         *slog.Logger
+}
+
+// DeviceData contains data related to a device endpoint.
+type DeviceData struct {
+	assetCreationTx chan AssetNotification
+	trackedAssets   map[string]AssetDeletionChan
+}
+
+// NewFileMountMap creates a new FileMountMap.
+func NewFileMountMap(fileMountPath string, logger *slog.Logger) *FileMountMap {
+	return &FileMountMap{
+		fileMountPath:  fileMountPath,
+		deviceMap:      make(map[string]DeviceData),
+		createDeviceTx: make(chan DeviceNotification, 100),
+		logger:         logger,
+	}
+}
+
+// UpdateDeviceEndpoints updates the device endpoints in the file mount map.
+func (f *FileMountMap) UpdateDeviceEndpoints(
+	devices map[string]DeviceEndpointRef,
+) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	for deviceKey := range f.deviceMap {
+		if _, exists := devices[deviceKey]; !exists {
+			if data, ok := f.deviceMap[deviceKey]; ok {
+				close(data.assetCreationTx)
+
+				for _, deletionChan := range data.trackedAssets {
+					close(deletionChan)
+				}
+			}
+			delete(f.deviceMap, deviceKey)
+		}
+	}
+
+	for deviceKey, deviceRef := range devices {
+		if _, exists := f.deviceMap[deviceKey]; !exists {
+			assetCreationTx := make(chan AssetNotification, 100)
+			assetCreationRx := make(chan AssetNotification, 100)
+
+			go func(tx, rx chan AssetNotification) {
+				for msg := range tx {
+					rx <- msg
+				}
+				close(rx)
+			}(assetCreationTx, assetCreationRx)
+
+			f.deviceMap[deviceKey] = DeviceData{
+				assetCreationTx: assetCreationTx,
+				trackedAssets:   make(map[string]AssetDeletionChan),
+			}
+
+			assetObserver := &AssetCreateObservation{
+				assetChan: assetCreationRx,
+			}
+
+			f.logger.Info("New device detected", "device", deviceRef.String())
+
+			f.createDeviceTx <- DeviceNotification{
+				DeviceEndpoint: deviceRef,
+				AssetObserver:  assetObserver,
+			}
+		}
+
+		f.updateAssets(deviceKey, deviceRef)
+	}
+}
+
+// updateAssets updates the assets associated with a device endpoint.
+func (f *FileMountMap) updateAssets(
+	deviceKey string,
+	deviceRef DeviceEndpointRef,
+) {
+	assets, err := getAssetNames(f.fileMountPath, deviceRef)
+	if err != nil {
+		f.logger.Warn(
+			"Failed to get asset names",
+			"error",
+			err,
+			"device",
+			deviceRef.String(),
+		)
+		return
+	}
+
+	deviceData, exists := f.deviceMap[deviceKey]
+	if !exists {
+		return
+	}
+
+	// Track assets that no longer exist
+	assetsToRemove := make([]string, 0)
+	for assetName := range deviceData.trackedAssets {
+		if _, exists := assets[assetName]; !exists {
+			assetsToRemove = append(assetsToRemove, assetName)
+		}
+	}
+
+	// Remove assets that no longer exist
+	for _, assetName := range assetsToRemove {
+		if deletionChan, exists := deviceData.trackedAssets[assetName]; exists {
```

Comment:
nit: It's slightly more idiomatic to use "ok" rather than "exists", even if the latter is more explicit.